man/man1/eins.1
.TH eins 1 "September 10, 2010" "SVN revision 5793" "USER COMMANDS" 
.SH NAME
eins \- timing and bandwidth measurement tool, especially for networks
.SH SYNOPSIS
.B eins
\-t type [MODOPTS] [\-i numtries] [\-n] [\-q] [\-u until] [\-b step by] host len
.SH DESCRIPTION
eins is a small network benchmarking util, written at the Institute of
Computer Science, University of Potsdam. It can measure latency/required time
of different tasks (network transfer, memory allocation) and determines the 
bandwidth. The measurement types are exposed by different modules. See module
section for details.
.PP
The basic idea of eins is to allow measurements with varying packet sizes.
Each measurement is repeated several times to improve the statistical quality. 
The measurement type, packet size and number of repetitions per packet size can 
be configured via command line arguments. See the following sections for details.

.SH OPTIONS

.TP
\-t type
use the given measurement module. Note, that this parameter has to be given
in front of MODOPTS 
.TP
MODOPTS
module specific options. See module section for details.
.TP
\-i numtries
Number of iterations used for each mesaurement (default is 54)
.TP
\-n 
be quiet, do not print results (useful for profiling) 
.TP
\-q
be more quiet, omit progress indicator
.TP
\-u limit
do the measurement up to the given limit starting from len paramater
.TP
\-b step
step the size of the measurement packets by the given step.
.TP 
host
specifies the measurement partner
.TP
len
The (initial) size of the measurement packets. If no limit is specified
only one measurement with the given size will be done.



.SH OUTPUT
eins outputs the measurement results in four columns with the following
content: 
.TP
size
The size of the measurement packet in Bytes
.TP
median
The median of all measured times (see numtries parameter) in microseconds.
See module section for how the times are determined.
.TP 
bandwith
This just the quotient of the first and the second column and
is therefore given in MByte/s
.TP
stddev
The standard deviation of all measured times.


.SH MODULES

.SS TCP measurement
This module basicly does a ping-pong measurement between a TCP server and
a client. Each measurement packet is sent to the server which immediatly
sends the packet back to the client. The time between the according send 
and recv system call on the client side is measured and devided by 2. Thus,
the reported time is equal to the half round-trip time (RTT) which can
be assumed to be the packet's latency. But be aware of this assumption!
.PP
To use the TCP module use 'tcp' as value for the type parameter. To start
a measurement a server instance of eins must be started. To do so, use
\-s as one of the module options. For further options see Common IP Options.

.SS UDP measurement
UDP measurement is done in a similar fashion as TCP. Sender and Receiver 
play a ping pong game with the measurement packets. The time between transmission
and reception of the packets is measured and devided by two (RTT) and gives
the resulting time of measurement.
.PP
To use the TCP module use 'udp' as value for the type parameter. Similar to
TCP, a server instance of eins must be started with the \-s option. A special
feature for UDP measurements is the fragmentation of measurement packets by
eins. By using the optional \-F option, followed by the desired fragment size
in Bytes, eins will split all packets by itself before sending them via
system call. For further options see Common IP Options.

.SS Common IP Options
The following Options are available for the measurement via IP networks. It is
safe to use them as MODOPTS (note that these have to occur after the type 
paramater)
.TP
\-6
Use Internet Protocol (IP) version 6. Version 4 is used by default
.TP
\-a address 
bind the socket to the given address (usefull when measurement should use
a specfic interface/network)
.TP
\-P port
Use the given port fot the sockets. A Server instance of eins will listen
on this port, while a client instance of eins will try to connect to this
port. The default value (if this parameter is not given) is 8910. 
.TP
\-H size
Prepend each packet with a header of the given size. Be aware of this 
feature: At the time of writing, the according code might be broken.

Please note, that for network measurement, a packet size (given by the 'len' 
parameter) of zero will cause a client to abort the measurement.

.SS Accept time measurement
With the accept module (usage: give 'accept' as value for the 'type' parameter)
measurement of accept time can be done. The measured time is the time that a
connect call on TCP client sockets takes. For such a measurement a TCP server
instance of eins has to be started (see TCP module for further details). This
module does not have any specific options.

.SS Memory management timings
This module allows the benchmarking of malloc and free call pairs. During each
measurement step a block of memory with the specified size is allocated via
malloc(3) syscall and disposed via free(3). The time that both calls take
is measured. This module does not have any specific options.

.SS Note on time measurement
Time measurement in eins is done by a call to clock_gettime. CLOCK_MONOTONIC
is used as reference clock. Therefore the measurements are not affected by
a system clock modification (caused by user or ntp, e.g.) or power management
features that lowers cpu clock rates. 

.SH EXAMPLES

.SS Simple TCP/UDP measurement example
Start a single TCP measurement on localhost with packet size of 1500.  
To do a UDP measurement, replace 'tcp' with 'udp'.

eins -t tcp -s &

eins -t tcp localhost 1500

.SS Extended TCP/UDP measurement example
A measurement via TCP with a starting packet size of 4 and upper limit of
2048 with a step of 2 bytes is done. Each measurement is repeated 100 times.
The Server should use port 1234 for communication. Both server and client use 
IPv6. The server is startet on a machine named example.domain

eins -t tcp -s -6 -P 1234 

eins -t tcp -6 -P 1234 -i 100 -b 2 -u 2048 example.domain 4

.SS Measure Malloc/Free timings
A straight forward example (note that you have to give the host option - 
althought it is ignored)

eins -t malloc -u 1024 localhost 1

.SH AUTHORS
Lars Schneidenbach and Hynek Schlawack (initial and most of eins code)
.PP
Joerg Jung (bugfixes and additional modules) and Steffen Christgau (bugfixes
and man page)


ChangeLog
2007-06-12  Lars Schneidenbach  <lschneid@cs.uni-potsdam.de>

	* src/mod_tcp.c (tcp_serve): handle ip-address option
	* src/mod_udp.c (udp_serve): handle ip-address option
	* src/mod_accept.c (accept_serve): handle ip-address option

	* src/util_ip (ip_prefs): included preferred ip-address as
	commandline parameter

2006-12-01  Lars Schneidenbach  <lschneid@cs.uni-potsdam.de>

	* src/mod_accept.c: new module to measure connect times

2006-10-26  Lars Schneidenbach  <lschneid@cs.uni-potsdam.de>

	* src/mod_tcp.c: Fragments included, TODO: put
	fragments code from tcp and udp together into ip!

2006-09-17  Hynek Schlawack  <hs+cl@ox.cx>

	* src/mod_malloc.c: Added basic memory benchmark.

	* src/mod_neon.c: Added support for NEON.

2006-06-27  Hynek Schlawack  <hs+cl@ox.cx>

	* src/mod_tcp.c (tcp_serve): Better handling of long messages.

2006-06-22  Hynek Schlawack  <hs+cl@ox.cx>

	* Global: Changed completely build system and uses
	autoconf/automake for building now.

2006-05-30  Hynek Schlawack  <hs+cl@ox.cx>

	* mod_bmi.c (bmi_measure): Added check whether the data received is
	the same a the sent.

2006-05-29  Hynek Schlawack  <hs+cl@ox.cx>

	* util_ip.c (ip_measure): Added check whether the data received is
	the same a the sent.
	(ip_measure): Replaced `int's with `size_t's in the signature.
	(ip_measure): Fixed a small introduced memleak.

2006-05-25  Lars Schneidenbach  <lschneid@cs.uni-potsdam.de>

	* AUTHORS: completed author list

2006-05-22  Hynek Schlawack  <hs+cl@ox.cx>

	* mod_bmi.c (bmi_serve): Fixed memleak.

	* mod_udp.c (udp_init): Sanitized fragmentation.

2006-05-15  Hynek Schlawack  <hs+cl@ox.cx>

	* eins.c (parse_args): Added check for valid `-i'.

	* mod_bmi.c (bmi_cleanup): Fixed memleak.

2006-05-12  Hynek Schlawack  <hs+cl@ox.cx>

	* mod_bmi.c (bmi_init): BMI_initialize() get run only the first
	time now.
	(bmi_cleanup): BMI_finalize() doesn't get called at all
	anymore. The reason for these changes is that BMI starts to behave
	strange when BMI_initialize()/BMI_finalize() get run too
	often (~1021).

	* eins.c (main): Enhaced progress indicator (added absolute value
	in bytes)

	* mod_bmi.c (bmi_init): Added code to handle immediate completion.

	* eins.c (main): Added support for iterating through several
	values.
	(main): Added simple progress indicator.
	(parse_args): Adjusted to new arguments.

	* eins.h (prefs): Added general prefs struct. Removed `no_time'
	from mod_args.

	* eins.c (GENERAL_USAGE): Clarified and enhaced usage.

	* ChangeLog: File added.


src/eins.c
/*
 * eins - A tool for benchmarking networks.
 * Copyright (C) 2006  Hynek Schlawack <hs+eins@ox.cx>
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License as
 * published by the Free Software Foundation; either version 2 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
 * 02110-1301, USA.
 */

// ISO
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>

// POSIX
#include <getopt.h>
#include <time.h>

// Own
#include "eins.h"
#include "measure.h"
#include "modules.h"
#include "util.h"

#define DEF_TRIES 54

#define GENERAL_USAGE \
    "Client: eins -t type [-i numtries] [-n] [-q] [-u until [-b step by]] host len\n" \
    "\t-t `type': Type of network we are benchmarking\n" \
    "\t-i `numtries': Number of iterations for measuring (default: 54)\n" \
    "\t-n: Don't print results (useful for profiling)\n" \
    "\t-q: Be quiet, omit progress indicator\n" \
    "\t-u `until': Iterate from `len' until the given value\n" \
    "\t-b `step by': When iterating using `-u', step by given value (default: 1)\n" \
    "\thost: Communicate with `host'\n" \
    "\tlen: Send packets of size `len'. If `-u' is used, start with this value.\n" \
    "\n"\
    "Server: eins -s -t type\n" \
    "\t-s: Start as server"

#define DEF_OPTS "st:i:nu:b:q"

static void
usage_and_die(void)
{
    puts(GENERAL_USAGE);

    puts("\nAvailable modules with options. The word in brackets is the type for the -t option."
         "\nThe -t option _must_ come _before_ any network specific options.");

    for (int i = 0; Modules[i]; i++) {
        printf("\n%s (%s):\n%s\n", Modules[i]->name, Modules[i]->type , Modules[i]->usage);
    }

    exit(1);
}

static char *
build_optstr(const net_mod *m[])
{
    size_t len = strlen(DEF_OPTS);
    for (int i = 0; m[i]; i++) {
        len += strlen(m[i]->opts);
    }

    char *str = safe_alloc(len + 1);
    strcpy(str, DEF_OPTS);

    for (int i = 0; m[i]; i++) {
        // Double entries seem to be okay as long they don't
        // contradict.
        strcat(str, m[i]->opts);
    }

    return str;
}

const net_mod *
parse_args(int argc, char * argv[], mod_args *ma, prefs *p)
{
    const net_mod *nm = NULL;

    char *optstr = build_optstr(Modules);

    char opt;
    while ((opt = getopt(argc, argv, optstr)) != -1) {
        // Mode
        switch (opt) {
        case 's':
            ma->mode = EINS_SERVER;
            break;

            // Mods
        case 't':
            for (int i = 0; Modules[i]; i++) {
                if (!strncmp(Modules[i]->type, optarg, strlen(optarg))) {
                    nm = Modules[i];
                }
            }
            break;

        case 'i':
            ma->tries = abs(atoi(optarg));
            if (ma->tries <= MIN_VALS)
                XL("Value of `-i' has to be bigger than %u.", MIN_VALS);
            break;

        case 'n':
            p->no_time = true;
            break;

        case 'u':
            p->until = atoi(optarg);
            break;

        case 'b':
            p->step = atoi(optarg);
            break;

        case 'q':
            p->quiet = true;
            break;

        default:
            if (nm) {
                if (!nm->handle_arg(opt, optarg)) {
                    L("Invalid option %c for `%s'.", opt, nm->name);
                    return NULL;
                }
            } else {
                L("No network type specified.");
                return NULL;
            }
        }
    }

    free(optstr); // Possible memleak but neglectible

    if (!nm) {
        usage_and_die();
    }

    if ((ma->mode == EINS_CLIENT) && ((argc != (optind + 2)))) {
        usage_and_die();
    }

    if (ma->mode == EINS_CLIENT) {
        ma->target = argv[optind];
        ma->size = atoi(argv[optind + 1]);
    }

    if (!p->until)
        p->until = ma->size;

    return nm;
}

/******************************************************************************
 * main()
 */
int
main(int argc, char **argv)
{
    mod_args ma = { EINS_CLIENT, NULL, DEF_TRIES, 0, NULL };
    prefs p = { 0, 1, false, false };
    const net_mod *nm;

    srand((unsigned int) time(NULL));

    if (!log_open(NULL)) {
        puts("Failed to open log file.");
        exit(1);
    }

    nm = parse_args(argc, argv, &ma, &p);
    if (!nm) exit(1);

    if (ma.mode == EINS_SERVER) {
        return !nm->serve(&ma);
    }

    // Set up time-measurement
    double *alltime, measuredelta = 0.0;
    time_586 ta, tb;

    init_timer();

    if (!p.no_time) {
        // Obtain time which is spent on measuring
        get_time(&ta);
        get_time(&tb);
        measuredelta = time_diff(tb, ta);
    }

    unsigned int progress = 0;
    for (; ma.size <= p.until; ma.size += p.step) {

        alltime = safe_alloc(ma.tries * sizeof(double));

        // Set up payload
        ma.payload = safe_alloc(ma.size);
        randomize_buffer(ma.payload, ma.size);

        // Set up mod
        if (!nm->init(&ma)) XL("Init/Handshake failed.");

        // Main measure-loop
        for (size_t try = 0; try < ma.tries; try++) {
            alltime[try] = (nm->measure() - measuredelta) / 2;
        }

        if (!p.no_time) {
            // Compute and print data
            double min, max, med, var;
            mean_variance(ma.tries, alltime, &min, &max, &med, &var);

	    // `bytes / 1000^-2 * s = bytes * 1000^2 / s',
	    // but we want `bytes * 1024^2 / s'
	    // 15625 / 16384 = (1000*1000) / (1024*1024)
            // Obsoleted by: Networks use Base10!! Lesson learned ;-)
            // => 1000B/s = 1kB/s; 1000kB/s = 1MB/s
	    double bw = ma.size / med;
	    //bw *= (double) 15625 / 16384;

            printf("%15d%16f%16f%16f\n", (int)ma.size, med, bw, sqrt(var));
        }

        nm->cleanup();

        free(alltime);
        free(ma.payload);

        unsigned int progress_new = ((double) ma.size / p.until) * 100;
        if (!p.quiet && progress_new > progress && !(progress_new == 100 && progress == 0 )) {
            progress = progress_new;
            L("%u%% (%u Bytes) done.", progress, ma.size);
        }
    }

    log_close();
}
